python/private/pypi_index.bzl

```diff
@@ -0,0 +1,302 @@
+
+    url = url if type(url) == type("") else url[0]
+
+    output_str = url
+    for char in [".", ":", "/", "\\", "$", "[", "]", "{", "}", "'", "\"", "-"]:
+        output_str = output_str.replace(char, "_")
+
+    output = ctx.path(output_str.strip("_").lower() + ".html")
+
+    real_url = envsubst(
+        url,
+        envsubst_vars,
+        ctx.getenv if hasattr(ctx, "getenv") else ctx.os.environ.get,
+    )
+
+    download = ctx.download(
+        url = [real_url],
+        output = output,
+        auth = get_auth(ctx, [real_url]),
+        **download_kwargs
+    )
+
+    return struct(
+        contents = lambda: _read_contents(
+            ctx,
+            download.wait() if download_kwargs.get("block") == False else download,
+            output,
+            url,
+        ),
+    )
+
+def _read_contents(ctx, result, output, url):
+    if not result.success:
+        fail("Failed to download from {}: {}".format(url, result))
+
+    html = ctx.read(output)
+    return get_packages(url, html)
+
+def get_packages_from_requirements(requirements_files):
+    """Get Simple API sources from a list of requirements files and merge them.
+
+    Args:
+        requirements_files(list[str]): A list of requirements files contents.
+
+    Returns:
+        A struct with `simpleapi` attribute that contains a dict of normalized package
+        name to a list of shas that we should index.
+    """
+    want_packages = {}
+    for contents in requirements_files:
+        parse_result = parse_requirements(contents)
+        for distribution, line in parse_result.requirements:
+            distribution = normalize_name(distribution)
+            shas = want_packages.get(distribution)
+            if not shas:
+                shas = sets.make()
+                want_packages[distribution] = shas
+
+            for sha in get_simpleapi_sources(line).shas:
+                sets.insert(shas, sha)
+
+    return struct(
+        simpleapi = want_packages,
+    )
+
+def get_simpleapi_sources(line):
+    """Get PyPI sources from a requirements.txt line.
+
+    We interpret the spec described in
+    https://pip.pypa.io/en/stable/reference/requirement-specifiers/#requirement-specifiers
+
+    Args:
+        line(str): The requirements.txt entry.
+
+    Returns:
+        A struct with shas attribute containing a list of shas to download from pypi_index.
+    """
+    head, _, maybe_hashes = line.partition(";")
+    _, _, version = head.partition("==")
+    version = version.partition(" ")[0].strip()
+
+    if "@" in head:
+        shas = []
+    else:
+        maybe_hashes = maybe_hashes or line
+        shas = [
+            sha.strip()
+            for sha in maybe_hashes.split("--hash=sha256:")[1:]
+        ]
+
+    if head == line:
+        head = line.partition("--hash=")[0].strip()
+    else:
+        head = head + ";" + maybe_hashes.partition("--hash=")[0].strip()
+
+    return struct(
+        wo_shas = line if not shas else head,
```

Comment:
Changed to just `requirement`, which is vague, but it might be good enough in the context of the struct and it makes it clearer to understand where it is used.